lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1711,6 +1747,118 @@ def convert_to_utf8(input)
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           },
+    #           "3": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": ...
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # The root level "code", "message", and "status" simply match the root
+    # cause of the first failed log entry. For example, if we switched the order
+    # of the log entries, then we would get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ...
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors that we care are:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    #
+    # The ultimate map that is constructed is:
+    # {
+    #   [7, 'User not authorized.']: ['0'],
+    #   [3, 'Log name contains illegal character :']: ['1', '3']
+    # }
+    def construct_error_details_map(error)
+      error_details_map = Hash.new { |h, k| h[k] = [] }
+
+      error_details = ensure_array(
+        ensure_hash(ensure_hash(JSON.parse(error.body))['error'])['details'])
+      partial_errors = error_details.detect(
+        -> { fail JSON::ParserError, "No type #{PARTIAL_ERROR_FIELD}." }
+      ) do |error_detail|
+        ensure_hash(error_detail)['@type'] == PARTIAL_ERROR_FIELD
+      end
+      log_entry_errors = ensure_hash(
+        ensure_hash(partial_errors)['logEntryErrors'])
+      log_entry_errors.each do |index, log_entry_error|
+        error_hash = ensure_hash(log_entry_error)
+        unless error_hash['code'] && error_hash['message']
+          @log.warn "Entry with index #{index} is missing 'code' or 'message'" \
+                    ' field.'
+          return {}
+        end
+        error_key = [error_hash['code'], error_hash['message']].freeze
+        # TODO(qingling128): Convert indexes to integers.
+        error_details_map[error_key] << index
+      end
+      error_details_map
+    rescue JSON::ParserError => e
+      @log.warn 'Failed to extract log entry errors from the error details:' \
+                " #{error.body}.", error: e
+    end
+
+    def ensure_array(value)
+      Array.try_convert(value) || (fail JSON::ParserError, "#{value.class}")
```

Comment:
[Really optional] This could also be:
```ruby
def check(cls, value)
  cls.try_convert(value) || (fail JSON::ParserError, "#{value.class}")
end
```
and then used as:
```ruby
      ...
      error_details = check(Array,
        check(Hash, check(Hash, JSON.parse(error.body))['error'])['details'])
      partial_errors = error_details.detect(
        -> { fail JSON::ParserError, "No type #{PARTIAL_ERROR_FIELD}." }
      ) do |error_detail|
        check(Hash, error_detail)['@type'] == PARTIAL_ERROR_FIELD
      end
      log_entry_errors = check(Hash,
        check(Hash, partial_errors)['logEntryErrors'])
      ...
```

Reply:
Hmm... I think this reduces readability in this case where we already have a lot of parentheses. Will keep the `ensure_array` and `ensure_hash` functions instead.

Reply:
The code I proposed has the exact same number of parentheses -- the only difference is that the `Array` and `Hash` bits are moved inside the parentheses. :-) And another advantage is that something like `check(String, index)` is also possible. But you're probably right that it's a bit too magical, so let's stay with what you have.